specs/dwritecore/DWriteCore.md

```diff
@@ -510,9 +510,189 @@ The following other DirectWrite APIs also use the hybrid font selection algorith
 -   `IDWriteTextLayout` if the typographic font family model is enabled by calling
     `IDWriteTextLayout4::SetFontAxisValues` or `IDWriteTextLayout4::SetAutomaticFontAxes`
 
-# Examples
-
-## Creating an isolated factory
+## Color Font Support
+
+By default, a glyph has a shape but no intrinsic color. Both DirectWrite and Direct2D have
+`DrawGlyphRun` methods that render glyph runs by filling the glyph shapes with a specified text
+color. For convenience, we will refer to this as _monochrome_ glyph rendering. All fonts have
+monochrome glyphs, but a color font also has color representations of some glyphs. To render glyphs
+in color, an application must use different glyph rendering APIs instead of calling the monochrome
+`DrawGlyphRun` method.
+
+Note that color font support does not affect text layout it at all. The text layout always produces
+monochrome glyph runs. Color is enabled at the glyph rendering level by substituting color
+representations for the monochrome glyphs.
+
+The easiest way to enable color is to call the `DrawGlyphRunWithColor` method instead of calling
+`DrawGyphRun`. Direct2D exposes this method through the `ID2D1DeviceContext7` interface. DirectWrite
+exposes this method through the `IDWriteBitmapRenderTarget3` interface.
+
+At a lower level, enabling color involves translating the monochrome _base glyph run_ to a sequence
+of _color glyph runs_. Each color glyph run must then be rendered using the appropriate method,
+depending on its _glyph image format_. The translation is performed by the factory object's
+`TranslateColorGlyphRun` method. It and related APIs are described in the following sections.
+
+### Using TranslateColorGlyphRun
+
+The `TranslateColorGlyphRun` factory method translates a monochrome _base glyph run_ into a sequence
+of _color glyph runs_. There are multiple ways of representiong color glyphs, known as _glyph image
+formats_ and identified by the `DWRITE_GLYPH_IMAGE_FORMATS` enumeration. When calling
+`TranslateColorGlyphRun`, the client specifies the combination of glyph image formats it supports.
+The `TranslateColorGlyphRun` returns a _glyph run enumerator_ object
+(`IDWriteColorGlyphRunEnumerator1` interface), which the client uses to iterate over the sequence of
+color glyph runs, represented by the `DWRITE_COLOR_GLYPH_RUN1` structure. Each color glyph run has a
+glyph image format. The client branches on the color run's glyph image format to determine how to
+render the color run.
+
+If the base glyph run does not have color glyphs in any of the requested glyph image formats, the
```

Comment:
What happens if the base glyph run contains some color glyphs that the client doesn't specify/support? Are those glyphs "missing" from the final iterator?

Reply:
Color glyphs in an unsupported format are treated the same as monochrome glyphs.

If there are _only_ monochrome glyphs (or only color glyphs in unsupported formats) then TranslateColorGlyphRun returns DWRITE_E_NOCOLOR, in which case the client should render the base glyph run.

If there's a mix of color and monochrome glyphs, then TranslateColorGlyphRun chops up the run. Color glyphs are translated to color runs as described in the overview section. Monochrome glyphs are returned as color runs with DWRITE_GLYPH_IMAGE_FORMATS_TRUETYPE or DWRITE_GLYPH_IMAGE_FORMATS_CFF and with paletteIndex = 0xFFFF. Then client then renders these as monochrome glyph runs using the foreground brush.

Reply:
When I read the text, I read that if no glyphs support any of the requested formats, you get back DWRITE_E_NOCOLOR. However I believe the text leaves the case ambiguous where only some of the glyphs are in the supported format. Can we try to clarify the text a bit? Specifically mentioning that the subset of glyphs which are not of the supported types will be "downconverted" to monochrome glyphs?

And on that note, if the downconversion takes place in that case where a subset of glyphs are supported, why do we need DWRITE_E_NOCOLOR at all? Why not just downconvert the entire set of glyph runs to monochrome glyphs, to eliminate the need for the client to handle that one error case?

Reply:
Returning DWRITE_E_NOCOLOR means we don't have to instantiate the enumerator object. Yes, I can better explain this case.

Reply:
I would argue that avoiding instantiating the enumerator object in this one-off, somewhat rare case, is not worth the additional API complexity of having to check for that special return value and then calling the monochrome method, every time this method is used. Especially true if conversion to monochrome already happens for color glyphs when their format is unsupported but other glyphs are supported. If what we gain from this is the cost of one allocation in a rare case, I think the reduced API complexity is preferable.

Reply:
This is existing API behavior going back to Windows 8.1. The thinking was that the client might call TranslateColorGlyphRun on every glyph run regardless of the font, and for a non-color font this would return DWRITE_E_NOCOLOR. Given this calling pattern, it's not a rare case.